tokio/Cargo.toml

```diff
@@ -59,6 +59,7 @@ net = [
   "windows-sys/Win32_System_SystemServices",
 ]
 process = [
+  "rt",
   "bytes",
   "libc",
   "mio/os-poll",
```

Comment:
It's awkward that we are introducing a dependency on the runtime here. This also means that we will be introducing additional error cases that we didn't have before, e.g. now the process is tied to a runtime and will stop working if the runtime goes away. 

Can we dynamically fall back to the signal implementation if the runtime goes away after creation?

Reply:
Actually the pidfd implementation does not depend on the runtime, it merely uses `PollEvented` which is present as long as feature `process` is enabled.

The feature `rt` is used in the `<PidfdReaper as Drop>::drop` implementation to spawn a new task to reap the children:

```
impl<W> Drop for PidfdReaper<W>
where
    W: Wait + Send + Sync + Unpin + 'static,
{
    fn drop(&mut self) {
        let mut reaper_inner = self.0.take().expect("inner has gone away");
        if let Ok(Some(_)) = reaper_inner.inner.try_wait() {
            return;
        }

        Handle {
            inner: reaper_inner.pidfd.scheduler_handle().clone(),
        }
        .spawn(async move {
            let _ = reaper_inner.await;
        });
    }
}
```

I uses `Handle::spawn` implementation since I don't want to write the same code twice.

Is there anyway to avoid use of `Handle` and spawn a task directly, while avoiding all the code duplication?

Reply:
Okay, but surely `PollEvented` does not work without rt even if it exists?

Reply:
Aha you are right xD, I can definitely add a dynamic fallback to signal.

Reply:
BTW how do I find out if a runtime has go away?
I thought tokio uses `sync::Weak<T>` or something like that for the handler but seems that it uses `Arc` in `scheduler::Handle`.

Reply:
To handle the case where the runtime has never been there: When you create the `PollEvented`, then use the `new_with_interest_and_handle` method, and to get the handle, you introduce a `Handle::try_current` in `src/runtime/scheduler/mod.rs` that returns an option instead of panicking when the handle is missing.

To handle the case where the runtime goes away: Trying to use the IO resource will return an error. Specifically, it returns the error created by calling `gone` from `src/runtime/io/registration.rs`. You can handle this error.

However, there's also a different possibility than this solution: We could spawn a thread with its own epoll instance for managing all pidfds in the process. That thread would have its own epoll instance and so on.

I think this probably requires some discussion about the best strategy.

Reply:
Thanks, btw can you provide me with a link to the Discord channel you use?

Reply:
It would be in #tokio-dev on https://discord.gg/tokio

Reply:
I've changed the `drop` code to use `GlobalOrphanQueue`, same as `Reaper`.